Add LLM example (#28)
CHTC/templates-GPUs: llm/README.md

# Personal CHTC Submit Template for LLM Fine-Tuning

> Use Case: Fine-tuning large language models on CHTC and track it with `Weights & Biases`.

Reviewer: It would help to know what the prepretained model is and what it is being fine tuned on. This doesn't need to be described in detail. You could link to the Hugging Face pages (e.g. https://huggingface.co/bert-base-cased).

We should add some sort of disclaimer that Weights & Biases is a popular third-party service for logging and visualizing model training but is not officially supported by CHTC. I like the idea of including it in this example (my group also uses it), but we don't want to have this come across as an official endorsement and users would have to troubleshoot W&B issues on their own.

![WANDB](wandb.png)

## Quick start

1. Store your WANDB credentials and `STAGING_DIR` path in an environment file named .env. See the provided [example](.env.example).

> 1. Update the `run_name` in `run.sh`. This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.

Reviewer: I believe we can move this to the submit file.

Reviewer: My suggested changes now move the run name to the submit file so the instructions can be modified to update that file with the run name and any other necessary changes. The script only needs to be modified to disable W&B logging.

Reviewer 2: Done in Christina's review suggestions.

1. Modify `run.sub` as necessary.
1. SSH to the submit node.
1. Create a `condor_log` directory using the command: `mkdir condor_log` if you don't have it.
1. Submit your job using `condor_submit run.sub`.

## Used Stacks

- Docker
- Github Container Registry (ghcr.io)
- Huggingface Transformers
- Weights & Biases (WANDB)

## Used CHTC/HTCondor Features

- Docker Universe
- Checkpointing

> - Staging (for storing checkpoints)

Reviewer: @ChristinaLK this template requires staging access. How should we document instructions for requesting that? Or do you have ideas to store checkpoints in a different way?

Reviewer 2: let's leave it for now...

- GPU

## Building your own container

Note: Perform this step on your local machine, not on a CHTC submit node.

Example resources for building a training container:

- [Dockerfile](Dockerfile)
- [requirements.txt](requirements.txt)
- [Helper script](build.sh)
- [.env](.env.example) for Github container registry credentials (`CR_PAT`)

> Users should consider building their own container to match their specific needs.

Reviewer: If this step is optional, we can move this to the top and explicitly note it as an optional step.

Reviewer 3: I should also mention that when I tested building my own container, a) it is a whopping 22GB and b) github defaulted to making it private so that my job couldn't fetch it. Is the token in `.env` supposed to be passed to condor somehow so it can fetch the ghcr container? Otherwise, we should mention the step of making sure the container is updated to public. I did verify that the token has `write/read: packages` set, so it should have the permission it needs to get the container.

@ChristinaLK As for the container being quite large, which is better -- using condor to move a large container over the network by pulling straight from ghcr.io, or should we set up `/staging/gpu-examples` like we did with `/squid/gpu-examples` for moving other large execution environments? (I assume 22GB is too large for squid to handle)

Reviewer: > it is a whopping 22GB

How did it get so big? The latest version of `huggingface/transformers-pytorch-gpu` has a compressed size of 10.38 GB on DockerHub. I would expect it already had all the large dependencies like CUDA, PyTorch, etc.

https://hub.docker.com/r/huggingface/transformers-pytorch-gpu/tags shows recent versions that are smaller, only 8 GB.

Author: Yes you are right, the [Dockerfile](https://github.com/JasonLo/templates-GPUs/blob/b9797b8572e029dd1e2ba0400d8d72aec07e4b49/llm/Dockerfile) only install a few small python dependencies on top of the original `huggingface/transformers-pytorch-gpu`. I had tried to rebuild it today, but the size is still around 22 GB... Maybe they are using some more aggressive compression? Feel free to swap my image to any CHTC provided one if you see fit.

Example Container Image:

- [Link](ghcr.io/jasonlo/chtc_condor:latest)

## FAQ

1. Why shouldn't I run python run.py directly in run.sub?

> I need to export the HuggingFace cache directory to _CONDOR_SCRATCH_DIR in a global scope. I'm unaware of a simple method to do this in python. Please let me know if you have a solution.

1. Why is `+GPUJobLength = "short"` present in `run.sub`?

> > The queuing duration for `long` is excessive, and since we perform checkpointing, it's more efficient to use `short`.

Reviewer: CHTC users are also allowed to run many more simultaneous short jobs than long jobs, so once checkpointing is supported short jobs are far better.

1. Can I use additional GPUs?

> > Absolutely! Just modify the `request_gpus` value in `run.sub` to your desired number.

Reviewer: Can we show where in `train.py` or indicate what package handles multi-GPU training? This is a mystery for many new GPU users.

Author: Added link to Huggingface's `Trainer`.

## To-Do List

- Consolidate all configurations into a single location? They are currently dispersed across `.env`, `run.sh`, and `run.sub`.
- Implement `wandb` hyperparameter `sweep` functionality.

> - Integrate `DeepSpeed` support.

Reviewer: I shared this example with my research group. One member recommended Hugging Face [Accelerate](https://huggingface.co/docs/accelerate/index) as an alternative way to train Hugging Face models with multiple GPUs. I don't suggest reworking this example, but perhaps it be listed as an alternative somewhere in the FAQ or TODO list.

Reviewer 2: Moved to issue #29

- Is it feasible or quicker to store the Docker image in `staging`?
- Experiment with a training-optimized container, such as [NGC](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch).

## About Me

> I am a data scientist at [DSI](https://datascience.wisc.edu/staff/lo-jason/).

Reviewer: Suggest rephrasing this to make it third-person, such as "About the Author" and "contributed by Jason Lo...".

Author: Done.

# Personal CHTC Submit Template for LLM Fine-Tuning

Use Case: Fine-tune large language models on CHTC and optionally monitor the process with Weights & Biases. This example is based on the example provided at [Hugging Face Documentation](https://huggingface.co/docs/transformers/training)

![WANDB](wandb.png)

## Quick start

1. Store your WANDB credentials and `STAGING_DIR` path in an environment file named .env. See the provided [example](.env.example). (If you do not have a CHTC `/staging` directory, contact the facilitation team at chtc@cs.wisc.edu, as described [here](https://chtc.cs.wisc.edu/uw-research-computing/file-avail-largedata)). 

> 1. Update the `run_name` in the submit file (in the `arguments = ` line). This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.

Reviewer: ```suggestion
1. Update the `run_name` in the submit file (in the `arguments = ` line). This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/results/run_name/`.
```

Example Container Image:

- [Link](https://github.com/users/jasonlo/packages/container/package/chtc_condor)

## Used Stacks

- Docker
- Github Container Registry (ghcr.io)
- Huggingface Transformers
- Weights & Biases (WANDB)

## Used CHTC/HTCondor Features

- Docker Universe
- Checkpointing
- Staging (for storing checkpoints)
- GPU

## FAQ

1. Why shouldn't I run python run.py directly in run.sub?

> I need to export the HuggingFace cache directory to _CONDOR_SCRATCH_DIR in a global scope. I'm unaware of a simple method to do this in python. Please let me know if you have a solution.

1. Why is `+GPUJobLength = "short"` present in `run.sub`?

> The queuing duration for `long` is excessive, and since we perform checkpointing, it's more efficient to use `short`. CHTC [policy](https://chtc.cs.wisc.edu/uw-research-computing/gpu-jobs) also allows users to run far more simultaneous `short` jobs than `long` jobs.

1. Can I use additional GPUs?

> Absolutely! Just modify the `request_gpus` value in `run.sub` to your desired number. HuggingFace's [trainer](https://huggingface.co/docs/transformers/main_classes/trainer) will then automatically use all available GPUs.

> 
> ## To-Do list

Reviewer: ```suggestion

1. How long does the model train?

> This example trains for a single epoch. In a research setting, you would modify the fine-tuning to train for more epochs or until training converges.

## To-Do list
```

# Personal CHTC Submit Template for LLM Fine-Tuning

Use Case: Fine-tune large language models on CHTC and optionally monitor the process with Weights & Biases. This example is based on the example provided at [Hugging Face Documentation](https://huggingface.co/docs/transformers/training)

![WANDB](wandb.png)

## Quick start

> 1. Store your WANDB credentials and `STAGING_DIR` path in an environment file named .env. See the provided [example](.env.example).

Reviewer 2: ```suggestion
1. Store your WANDB credentials and `STAGING_DIR` path in an environment file named .env. See the provided [example](.env.example). (If you do not have a CHTC `/staging` directory, contact the facilitation team at chtc@cs.wisc.edu, as described [here](https://chtc.cs.wisc.edu/uw-research-computing/file-avail-largedata)). 
```

> 1. Update the `run_name` in `run.sh`. This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.

Reviewer 2: ```suggestion
1. Update the `run_name` in the submit file. This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.
```

> 1. Update the `run_name` in the submit file. This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.

Reviewer 2: ```suggestion
1. Update the `run_name` in the submit file (in the `arguments = ` line). This will be utilized as the WANDB tracking ID, and checkpoints will be saved in `STAGING_DIR/run_name/`.
```

1. (Optional) Build your own training container, see details below.
1. Modify `run.sub` as necessary.

> 1. SSH to the submit node.

Reviewer 2: ```suggestion

```

1. Create a `condor_log` directory using the command: `mkdir condor_log` if you don't have it.
1. Submit your job using `condor_submit run.sub`.

## Building your own container (Optional)

Note: Perform this step on your local machine, not on a CHTC submit node.

Example resources for building a training container:

- [Dockerfile](Dockerfile)
- [requirements.txt](requirements.txt)
- [Helper script](build_push_container.sh)
- [.env](.env.example)

Users should consider building their own container to match their specific needs.

Example Container Image:

- [Link](https://github.com/users/jasonlo/packages/container/package/chtc_condor)

## Used Stacks

- Docker
- Github Container Registry (ghcr.io)
- Huggingface Transformers
- Weights & Biases (WANDB)

## Used CHTC/HTCondor Features

- Docker Universe
- Checkpointing
- Staging (for storing checkpoints)
- GPU

## FAQ

1. Why shouldn't I run python run.py directly in run.sub?

> I need to export the HuggingFace cache directory to _CONDOR_SCRATCH_DIR in a global scope. I'm unaware of a simple method to do this in python. Please let me know if you have a solution.

1. Why is `+GPUJobLength = "short"` present in `run.sub`?

> > The queuing duration for `long` is excessive, and since we perform checkpointing, it's more efficient to use `short`.

Reviewer: ```suggestion
> The queuing duration for `long` is excessive, and since we perform checkpointing, it's more efficient to use `short`. CHTC [policy](https://chtc.cs.wisc.edu/uw-research-computing/gpu-jobs) also allows users to run far more simultaneous `short` jobs than `long` jobs.
```
